vmdb/app/controllers/api_controller/providers.rb

```diff
@@ -0,0 +1,126 @@
+class ApiController
+  module Providers
+    CREDENTIAL_ATTRS = %w(userid password)
+    RESTRICTED_ATTRS = %w(type) + CREDENTIAL_ATTRS
+
+    def create_resource_providers(type, _id, data = {})
+      if data.key?("id") || data.key?("href")
+        raise BadRequestError,
+              "Resource id or href should not be specified for creating a new #{type}"
+      end
+
+      create_provider(data)
+    end
+
+    def edit_resource_providers(type, id = nil, data = {})
+      raise BadRequestError, "Must specify an id for editing a #{type} resource" unless id
+      raise BadRequestError, "Provider type cannot be updated" if data.key?("type")
+
+      klass = collection_config[:providers][:klass].constantize
+      provider = resource_search(id, type, klass)
+      edit_provider(provider, data)
+    end
+
+    def refresh_resource_providers(type, id = nil, _data = nil)
+      raise BadRequestError, "Must specify an id for refreshing a #{type} resource" unless id
+
+      api_action(type, id) do |klass|
+        provider = resource_search(id, type, klass)
+        api_log_info("Refreshing #{provider_ident(provider)}")
+
+        refresh_provider(provider)
+      end
+    end
+
+    def delete_resource_providers(type, id = nil, _data = nil)
+      raise BadRequestError, "Must specify an id for deleting a #{type} resource" unless id
+
+      api_action(type, id) do |klass|
+        provider = resource_search(id, type, klass)
+        api_log_info("Deleting #{provider_ident(provider)}")
+
+        destroy_provider(provider)
+      end
+    end
+
+    private
+
+    def provider_ident(provider)
+      "Provider id:#{provider.id} name:'#{provider.name}'"
+    end
+
+    def fetch_provider_klass(klass, data)
+      supported_types = klass.supported_subclasses.collect(&:name)
+      types_string    = supported_types.join(", ")
+
+      unless data.key?("type")
+        raise BadRequestError, "Must specify a provider type, supported types are: #{types_string}"
+      end
+      type = data["type"]
+      unless supported_types.include?(type)
+        raise BadRequestError, "Invalid provider type #{type} specified, supported types are: #{types_string}"
+      end
+      klass.supported_subclasses.detect { |p| p.name == data["type"] }
```

Comment:
I think this has to be tweaked slightly for when you are creating/editing a Foreman provider, which is added via ProviderForeman vs Vmware which is added via EmsVmware.  There is an inconsistency until we get all of the providers actually using the providers table.

Note that for Foreman, you only create a ProviderForeman, and it handles creating the managers, so there only one thing to enter.  For Openstack, you will eventually have *two* things, a cloud side and an infra side.  I think this needs to be taken into account in the design, but not necessarily this PR.  I just think we should think it out so that the API can stay consistent as we move forward without messing up the API consumers with lots of changes.